src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java

```diff
@@ -78,23 +92,24 @@ public List<BufferLease> take(long size) {
     }
 
     public void offer(BufferLease bufferLease) {
-        if (bufferLease.attemptRelease()) {
-            internalOffer(bufferLease);
-        }
+        bufferLease.removeRef();
     }
 
-    private void internalOffer(BufferLease bufferLease) {
-        if (!bufferLease.isStub()) {
-            queue.add(bufferLease);
+    void internalOffer(BufferContainer bufferContainer) {
+        // Adding back to pool:
+        // - If stub, add container stub to queue
+        // - If not, add container from lease
+        if (!bufferContainer.isStub()) {
+            queue.add(bufferContainer);
         }
 
         if (close.get()) {
             LOGGER.debug("closing in offer");
             while (queue.peek() != null) {
                 if (lock.tryLock()) {
                     while (true) {
-                        BufferLease queuedBufferLease = queue.poll();
-                        if (queuedBufferLease == null) {
+                        BufferContainer queuedBufferContainer = queue.poll();
+                        if (queuedBufferContainer == null) {
                             break;
                         }
                     }
```

Comment:
Will this implementation consume all available processing power in a busy loop until poll returns null?